File: cpp/misc/Adt.cpp
/** @file Adt.cpp 
 *
 *  @author Jon Turner
 *  @date 2012
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

#include "Adt.h"

namespace grafalgo {

/** Read an index from the input stream.
 *  An index typically represents some component of a data structure,
 *  such as an element of a set or vertex in a graph.
 *  By convention, data structures whose index sets have at most 26 elements
 *  have a string representation that substitutes lower-case
 *  letters for the index values used internally (so 1 becomes 'a',
 *  2 becomes 'b' and so forth). On input, if the next non-space
 *  character is a lower-case letter, we replace 'a' with 1, etc.
 *  If the next non-space character is a digit, we read an integer
 *  and interpret it as an index.
 *  @param in is an open input stream
 *  @param x is a reference to an index in which the input value is
 *  returned
 *  @return true on success, else false
 */
bool Adt::readIndex(istream& in, index& x) {
	Util::skipSpace(in);
	char c = in.peek();
	if (!in.good()) return false;
	if (islower(c)) {
		x = (c - 'a') + 1; c = in.get(); return in.good();
	} else if (isdigit(c)) {
		in >> x; return in.good();
	}
	return false;
}

/** Convert an index to a string.
 *  @param[in] x is a valid index for the data structure
 *  @return a string that represents the value of x; if n()>26, this is
 *  just the string representing the number x, otherwise, it is a lower-case
 *  letter.
 */
string Adt::index2string(index x) const {
	string s = "";
	if (n() <= 26) {
		if (x == 0) s += '-';
		else s += ((char) ((x-1) + 'a'));
	}
	else s += to_string(x);
	return s;
}

} // ends namespace

File: js/dataStructures/Adt.mjs
/** @file Adt.java
 *
 *  @author Jon Turner
 *  @date 2021
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

/** The Adt class is a base class from which other data structures
 *  in grafalgo are derived.
 *
 *  The data structures in grafalgo are built using integer index values
 *  to refer to specific items (set elements, nodes in search trees,
 *  vertices in graphs). In this context, an index is a positive integer
 *  in a bounded range 0..n, for some value of n (0 is used as null).
 *
 *  The use of index values has a couple advantages over pointers.
 *  First, index values can serve as common "handles" for
 *  items in multiple data structures, eliminating the need to have
 *  explicit mappings to relate such items. As one example,
 *  in an algorithm that uses a graph and a separate list of vertices,
 *  both can use the same indexes to represent the vertices.
 *  Index values also make it trivial to have fast membership tests
 *  for index lists and similar data structures.
 */
export default class Adt {
	#n;		// index values in 0..n

	constructor(n=26) { this.#n = n; }

	/** Get the index range for the object.
	 *  @return the largest index value
	 */
	get n() { return this.#n; }

	/** Set the index range for the object.
	 *  For use of subclasses only.
	 *  @param nn becomes the largest index value
	 */
	set _n(n) { this.#n = n; }

	/** Determine if a given index is valid.
	 *  @param[in] i is an integer
	 *  @return true if i lies within the allowed range of index values,
	 *  else false
	 */
	valid(i) { i = Math.floor(i); return 0 <= i && i <= this.#n; }

	/** Create a string representation of the data structure.
	 *  Should show all information relevant to the provided abstraction.
	 *  Subclasses are all expected to implement this method.
 	 */
	toString() { return ""; };

	/** Compare two Adts for equality.
	 *  @param a2 is an Adt object to be compared to this.
	 *  @return true if this is equal to a2; for subclasses that do not
	 *  provide their own equality check, comparison defaults to equality
	 *  of the string representations.
	 */
	equals(a2) { return this.toString() == a2.toString(); }

	/** Compare two Adts for equality, including internal details.
	 *  @param a2 is an Adt object to be compared to this.
	 *  @return true if this is equal to a2, not just at the abstraction
	 *  level, but also for internal details; for subclasses that do not
	 *  provide their own long equality check, comparison defaults to equality
	 *  of the long string representations.
	longEquals(a2) {
		return (typeof a2) == 'string' ?
					this.toLongString() == a2 :
					this.toLongString() == a2.toLongString();
	}
	 */

	/** Check the data structure for internal consistency.
	 *  @return true if the data structure is internally consistent,
	 *  else return false; if a subclass does not supply a consistent()
	 *  method, false is returned.
	consistent() { return false; }
	 */

	/** Convert an index to a string.
	 *  @param[in] x is a valid index for the data structure
	 *  @return a string that represents the value of x; if this.n>26, this is
	 *  just the string representing the number x, otherwise, it is a lower-case
	 *  letter.
	 */
	index2string(x, strict=false) {
		const alpha = '-abcdefghijklmnopqrstuvwxyz';
		let s = "";
		if (!strict && 0 <= this.n && this.n <= 26) {
			s += alpha[x];
		} else {
			s += x;
		}
		return s;
	}

}
